src/receivers/kafka/offset_ack_committer.rs

```diff
@@ -0,0 +1,228 @@
+                            );
+                            ack_count += 1;
+                        }
+                        payload::KafkaAcknowledgement::Nack(_) => {
+                            // TODO: Implement nack handling
+                        }
+                    }
+                }
+                None => {
+                    debug!("Ack channel closed during drain after {} acks", ack_count);
+                    break;
+                }
+            }
+        }
+
+        if ack_count > 0 {
+            debug!("Drained {} pending acknowledgements", ack_count);
+        }
+
+        // Perform the final commit using the existing commit_offset function
+        debug!("Performing final commit after draining");
+        commit_offset(
+            self.assigned_partitions.clone(),
+            &self.topic_names_to_id,
+            &self.topic_trackers,
+            |tpl, mode| self.consumer.commit(tpl, mode),
+        );
+    }
+}
+
+pub fn commit_offset<F>(
+    assigned_partitions: Arc<std::sync::Mutex<AssignedPartitions>>,
+    topic_name_to_id: &HashMap<String, u8>,
+    topic_trackers: &Arc<TopicTrackers>,
+    commit_fn: F,
+) where
+    F: FnOnce(
+        &rdkafka::topic_partition_list::TopicPartitionList,
+        CommitMode,
+    ) -> rdkafka::error::KafkaResult<()>,
+{
+    // Lock and clone assigned partitions
+    let assigned = assigned_partitions.lock().unwrap().clone();
+
+    // Build list of topic/partition/offset to commit
+    let mut commits = Vec::new();
+    for (topic_name, partitions) in assigned.topics.iter() {
+        let topic_id = match topic_name_to_id.get(topic_name) {
+            Some(id) => *id,
+            None => {
+                debug!("Unknown topic name in assigned partitions: {}", topic_name);
+                continue;
+            }
+        };
+
+        // For each assigned partition, determine the offset to commit
+        for &partition in partitions.iter() {
+            // Try to get lowest pending offset first
+            if let Some(offset) = topic_trackers.lowest_pending_offset(topic_id, partition) {
+                debug!(
+                    "Topic {} (id {}) partition {} has pending offset: {}",
+                    topic_name, topic_id, partition, offset
+                );
+                commits.push((topic_name, partition, offset));
+            } else if let Some(hwm) = topic_trackers.high_water_mark(topic_id, partition) {
+                debug!(
+                    "Topic {} (id {}) partition {} has no pending offsets, using high water mark: {}",
+                    topic_name, topic_id, partition, hwm
+                );
+                commits.push((topic_name, partition, hwm + 1));
+            } else {
+                debug!(
+                    "Topic {} (id {}) partition {} has no pending offsets and no high water mark",
+                    topic_name, topic_id, partition
+                );
+            }
+        }
+    }
```

Comment:
It wasn't clear, but do we want to continue holding the `assigned` lock past here or could it be dropped early?